include/motor.hpp

```diff
@@ -0,0 +1,116 @@
+#ifndef MOTOR_H
+#define MOTOR_H
+
+#include "Arduino.h"
+#include "freertos/FreeRTOS.h"
+#include "freertos/task.h"
+#include "esp_log.h"
+//#include "freertos/semphr.h"
+
+/**
+ *  Motor
+ *   
+ *  position = mm
+ *  speed = mm/s
+ *  acceleration = mm/s^2
+ */
+
+#define DEBUG_CLIPPING              // Show debug messages when motions violating the machine 
+                                    // physics are commanded
+
+typedef struct {
+  float start; // Should always be 0
+  float end;
+  float keepout; /*  Soft endstop preventing hard crashes in mm. Will be 
```

Comment:
We should re-define the coordinate space to start at 0, and not use negative numbers for the home position. initially 0 was at the keepout position because it made patterns simple. If we are having patterns not work in the main coordinate space, but just work in a translated space, this no longer is an issue

Reply:
I would still prefer to stick with the coordinate system used currently. The safe gotopos() should constrain the coordinate system to start / end. It should not allow movement into negative. The coordinate system presented to the user should start at 0. I see no issues with the endstop at a negative position. Btw. This is not uncommon in CNC machines.

Reply:
So should we keep the idea of the Motion Bounds separate from the Motor Coordinate Space?
The Motion Bounds doesn't describe the true start/end of the Coordinate Space, but that of the Machine itself.
For example, my LinMot has a Motion Bounds of `{.start = 110mm, .end = 0mm, .keepout = 5mm}`. Thats describing the actual space the real machine lives in.

Lets say the user has a machine with 100mm of reach, a home switch at 0mm, and they wanted 5 mm of keep out.
I'd want the user to be able to specify that as simply as possible, and have the internals figure out the acceptable range.

I think a user would have a harder time if they are being asked to specify `motor->setMotionBounds({ .start = 0, .end = 90, .keepout = 5 })`

rather than

`motor->setMotionBounds({ .start = 0, .end = 100, .keepout = 5 })`

I feel like its gonna cause confusion if we are asking the user to subtract the `.keepout` from `.end`, when the internal motor could easily do it for them

So the user specifies the Motion Bounds for the machine, and the Motor itself understands position 0 in our coordinate space corresponds to 5mm in the real machine. That makes it simpler for the user, right?

Reply:
Maybe we need to rename it from Motion Bounds to something better? Maybe the Motion Bounds really are the limits of the coordinate space, and would have negatives be where home really is. Maybe this is a problem because what I'm specifying here really is the Machine Bounds

Reply:
Also, to clarify, LinMot is troublesome in that it forces the coordinate system + direction to be aligned to a specific fixed direction of the motor, so if you want to run it the other way, you need to use a reverse coordinate system. It's somewhat annoying, but workaround-able

Reply:
So yeah, I think what I'm saying is we should have an easy API for the user to configure the machine, but the internal motion bounds will still use the system you are describing. The true home would still be at like -5mm for example, because keepout would still be in the negative area.

I just don't want the user to have to compute that, or worry about that. I just want them to get a measuring tape and pop the numbers in

Reply:
So when the user provides the Machine Bounds, we compute the Motion Bounds from them, where keepouts would be negative

Reply:
Especially because the User won't actually ever see the real Motor Coordinate Space. They'll always be operating in the Pattern Coordinate Space which is translated by the StrokeEngine

Reply:
So I'm thinking of replacing setBounds with `virtual void setMachineLimits()`. So LinMots can set Machine Limits using start, end, keepout. Steppers can set Machine Limits using length, and keepout. That keeps setup minimal, and friendly to the user

We'll compute our own Motion Bounds internally which won't be exposed to the user for constraining motion commands

Reply:
or maybe I'll have it not virtual, and still just use the same start/end/keepout format so we can keep it consistent between implementations

Reply:
We'd end up having three Coordinate Spaces
- Machine Coordinate Space (Really only applicable for motors which send Motion Commands via ModBus/CANOpen/EtherCat)
- Bounded Coordinate Space (0 to length, with -keepout, and length + keepout zones)
- Pattern/Stroke Coordinate Space (0 to stroke, translated around Bounded Coordinate Space by StrokeEngine via Depth parameter)

Reply:
Machine Coordinate Space and Bounded Coordinate Space would end up being the same for Stepper Motors which lack the global positioning system of an external driver with built-in homing (like Linmot)

Reply:
`goToPos` would operate within the Bounded Coordinate Space, and would need to be mapped to the Machine Coordinate Space by the individual driver implementation

Reply:
For Stepper Drivers, there would be no mapping needed

Reply:
I pushed some updated code to showcase what I mean. This is how I'm using it in the LinMot driver as well
```
motor = new LinmotMotor();

machineLimits bounds = {
  .start = 110, // mm
  .end = -7, // mm
  .keepout = 10 // mm
};
motor->setMaxSpeed(5000); // 5 m/s
motor->setMaxAcceleration(25000); // 25 m/s^2
motor->setMachineLimits(bounds);
```

```
void LinmotMotor::unsafeGoToPos(float position, float speed, float acceleration) {
  if (!this->isInState(MotorState::ACTIVE)) {
    ESP_LOGE("LinmotMotor", "Attempted to issue Motion CMD while in incorrect state '%s'!", this->getStateString());
    return;
  }

  this->CO_sendCmd(
    0x0900, 
    static_cast<uint16_t>(this->mapSafePosition(position) * 10), 
    static_cast<uint16_t>(speed), 
    static_cast<uint16_t>(acceleration), 
    static_cast<uint16_t>(acceleration)
  );
}
```

Reply:
So we provide two forms of protection. That in the safe `goToPos`, and the protected `mapSafePosition`. both will ensure the Machine Limits are followed, and we don't attempt to command anything outside of them

Reply:
In the end we have three coordinate systems. The one the machine lives in, the bounded one StrokeEngine lives in with the keepout area applied. And finally the pattern coordinate system which is offset by depth.

May I ask you to stick with the terms I've layout for StrokeEngine. That might have prevented some confusion on my site. 

But yes, we should separate the machine's coordinate system from the one StrokeEngine is using. Your proposed ` void setMachineLimits()` seems a good idea. The motor driver is then responsible to create the bounded coordinate system for StrokeEninge and translate between the two. This is how I had layout it out initially. Maybe we even stick with the original naming / struct:
```
static machineGeometry strokingMachine = {
  .physicalTravel = 160.0,            // Real physical travel from one hard endstop to the other
  .keepoutBoundary = 5.0              // Safe distance the motion is constrained to avoiding crashes
};
```

Reply:
Sorry about the confusion with the naming there. I should have been more clear than just calling it Motion Bounds.
I do think making a Terminology markdown file would be useful so we can reduce confusion like this in the future.

If I have some time in the future, I'll try to make a Terminology file.

Reply:
In the end I plan for the iHSV57 to use sensorless homing to measure the actual physical travel.